### PR TITLE
Improve error when checksum dict has no entry for a file

### LINKED ISSUE #4254: New checksum listing with dictionaries leads to confusing errors if you forget to update them
From a PR which had forgotten to inject new checksums
```python
name = 'VTK'
version = '9.2.6'
...
checksums = [
    {'VTK-9.2.2.tar.gz': '1c5b0a2be71fac96ff4831af69e350f7a0ea3168981f790c000709dcf9121075'},
    {'VTKData-9.2.2.tar.gz': '4b0ac438c77773ff741fd02c0e3b0dc1406b4c51e6cbe4e8304306fa125f27bf'},
    {'vtk-version.egg-info': '787b82415ae7a4a1f815b4db0e25f7abc809a05fc85d7d219627f3a7e5d3867b'},
]
```
with the problem being that they are mapping to the old versions (and, of course, the checksums being old as well.
They are technically correct for version 9.2.2 (so there isn't actually any errors here if one were to nitpick)

but it does lead to a very confusing error message
```
== 2023-05-05 17:55:58,351 build_log.py:171 ERROR EasyBuild crashed with an error (at easybuild/easybuild-framework/easybuild/base/exceptions.py:126 in __init__): Invalid checksum spec 'None', should be a string (MD5) or 2-tuple (type, value). (at easybuild/easybuild-framework/easybuild/tools/filetools.py:1306 in verify_checksum)
```

Invalid checksum spec 'None'? It should detect the lack of a checksum for the particular source and give a proper error for that specifically.

### PR DESCRIPTION
**Pull request by Flamefire**
When the dict didn't contain the filename EB will crash with
> Invalid checksum spec 'None', should be a string (MD5) or 2-tuple (type, value).

This will now raise a more descriptive error.

I'm not sure how this should be handled. IMO an error is the right solution especially as the recursive call will consider it also as a found valid checksum even when all others failed.

But not sure if this is correct but it is at least backwards compatible: It just improves the error message to e.g.: "Missing checksum for test.txt in {'wrong-name': '1c49562c4b404f3120a3fa0926c8d09c99ef80e470f7de03ffdfa14047960ea5'}"

The alternatives all have strange edge cases

See also https://github.com/easybuilders/easybuild-docs/pull/104

Fixes #4254

### CODE REVIEW

**Comment by Flamefire** [comment]
Rebased again although I'd like to point to https://github.com/easybuilders/easybuild-docs/pull/104 again where I put up a proposed documentation for how checksums should be handled (and partially(!) are right now) which should be discussed and decided on as otherwise it is hard to "fix" things when you don't know what the correct behavior should be.

E.g. here I assume that each possible filename must be listed in the dict, i.e. a missing filename is an error and NOT equivalent to a filename key with `None` value.    
Note that `None` as the value is still not handled (see #4142) as that requires defining **how** it should be handled (e.g. via recursive call to get to https://github.com/easybuilders/easybuild-framework/pull/4150/files#diff-0ca2799267ffd6a801a47e695fbe66c047a44775b616b26148100ec11e6ce5e5R1256) especially related to checksum alternatives: Currently an alternative of `None` would trigger an exception when checksums are enforced even if another (later) alternative would match.

**Review by boegel — CHANGES_REQUESTED** [review]

File: easybuild/tools/filetools.py [hunk 1]
```diff
@@ -1265,14 +1265,11 @@ def verify_checksum(path, checksums):
...
+            try:
                 # Set this to a string-type checksum
                 checksum = checksum[filename]
-            elif build_option('enforce_checksums'):
-                raise EasyBuildError("Missing checksum for %s", filename)
-            else:
-                # Set to None and allow to fail elsewhere
-                checksum = None
+            except KeyError:
+                raise EasyBuildError("Missing checksum for %s in %s", filename, checksum)
```
[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/tools/filetools.py:1272]: I wanted to keep the current behavior that this situation is an error and just provide a meaningful error message.

I drafted https://github.com/easybuilders/easybuild-docs/pull/104 where there are some open question on how checksums should be handled which is likely a breaking change (or a handful of breaking changes)

IMO a dict with a missing filename/key is an error: Likely the dict was meant for arch-specific files&checksums but a new arch is used which wasn't tested yet. Or a spelling error happened making it always fail to verify checksums even if it should.    
That's what I suggested in the draft mentioned above: A dict, if present, needs to list each possible filename even without `enforce-checksums`

[the remark this answers is not in the document]
Comment (Flamefire) [reply easybuild/tools/filetools.py:1272]: BTW: With a [recent issue](https://github.com/easybuilders/easybuild-easyconfigs/pull/18491#issuecomment-1670182732) I noticed that `enforce-checksums` might be to greedy: We **need** a `None` value if the archive (e.g. Git checkout) isn't checksummable. But we want to check that all other checksums are present.

Now we don't (usually) have lists anymore but lists of dicts of filename-to-checksum. Now if `enforce-checksums` triggers for every explicit `None` but a missing filename in a dict doesn't trigger an error if it isn't set we have no way to check e.g. the (old) PyTorch ECs reliably.

Hence I think the approach I suggested makes sense.

[the remark this answers is not in the document]
Comment (boegel) [reply easybuild/tools/filetools.py:1272]: Thanks for clarifying, I agree that this is at least a good step forward to avoid totally meaningless errors.

If you noticed any other problems, please make sure there's an issue open for it (or if not, open one)

**Review by boegel — APPROVED** [review]
